temporal-opentracing/src/main/java/io/temporal/opentracing/internal/OpenTracingWorkflowInboundCallsInterceptor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ *  Copyright (C) 2020 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.opentracing.internal;
+
+import io.opentracing.Scope;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.util.GlobalTracer;
+import io.temporal.common.interceptors.WorkflowInboundCallsInterceptor;
+import io.temporal.common.interceptors.WorkflowInboundCallsInterceptorBase;
+import io.temporal.common.interceptors.WorkflowOutboundCallsInterceptor;
+import io.temporal.opentracing.OpenTracingOptions;
+import io.temporal.opentracing.SpanOperationType;
+import io.temporal.workflow.Workflow;
+
+public class OpenTracingWorkflowInboundCallsInterceptor
+    extends WorkflowInboundCallsInterceptorBase {
+  private final OpenTracingOptions openTracingOptions;
+
+  public OpenTracingWorkflowInboundCallsInterceptor(
+      WorkflowInboundCallsInterceptor next, OpenTracingOptions openTracingOptions) {
+    super(next);
+    this.openTracingOptions = openTracingOptions;
+  }
+
+  @Override
+  public void init(WorkflowOutboundCallsInterceptor outboundCalls) {
+    super.init(new OpenTracingWorkflowOutboundCallsInterceptor(outboundCalls, openTracingOptions));
+  }
+
+  @Override
+  public WorkflowOutput execute(WorkflowInput input) {
+    Tracer tracer = GlobalTracer.get();
+    SpanContext rootSpanContext =
+        OpenTracingContextAccessor.readSpanContextFromHeader(input.getHeader(), tracer);
+    Span workflowRunSpan =
+        OpenTracingSpanUtils.createWorkflowRunSpan(
+                tracer,
+                Workflow.currentTimeMillis(),
+                openTracingOptions.getSpanOperationNamePrefix(SpanOperationType.WORKFLOW_RUN)
+                    + "-"
+                    + Workflow.getInfo().getWorkflowType(),
+                Workflow.getInfo().getWorkflowId(),
+                Workflow.getInfo().getRunId(),
+                rootSpanContext)
+            .start();
+    try (Scope scope = tracer.scopeManager().activate(workflowRunSpan)) {
+      return super.execute(input);
+    } catch (Throwable t) {
+      OpenTracingSpanUtils.logFail(workflowRunSpan, t);
+      throw t;
+    } finally {
+      workflowRunSpan.finish();
```

Comment:
This is going to close the workflow span every time a workflow is unloaded from the worker cache. I think it is pretty confusing to see many spans instead of one span for a workflow. I propose the following solution to this:

Any span created inside the workflow should be created inside a `SideEffect` function. It should be serialized using `Tracer.inject` as SideEffect requires its result to be serializable as it is stored as a _marker_ event in the event history. Then it has to be deserialized using `Tracer.extract` before use. 

`workflowRunSpan.finish()` should not be called on exceptions that extend Error, especially `DestroyWorkflowThreadError` which is used to release threads on workflow unloading from memory. The interceptor interface should be extended to give a clear indication if an exception is going to fail workflow or workflow task only.

Reply:
What do you think about making this a separate change and PR and implement it as an improvement for the current solution?
The current approach is simple, it works and it could be good enough for the majority of cases.

Also, this idea looks very brittle for me and I think it will not work. The first problem that I see is actually in the root of the proposal:
> This is going to close the workflow span every time a workflow is unloaded from the worker cache.
> workflowRunSpan.finish() should not be called on exceptions that extend Error, especially DestroyWorkflowThreadError which is used to release threads on workflow unloading from memory.

We should call finish in all these cases. Otherwise, imagine a workflow started today and finishing a month from now. A workflow run span of this workflow will not be finished till a month from now? It will not be reported to the server in that case and we will not have any trace.

We should actually wrap every "awakening" as a separate span to keep spans published without huge delays.

Reply:
OK. Let's keep your approach. I didn't realize that children of non-closed span are not reported. So it is going to be useless for long running workflows.